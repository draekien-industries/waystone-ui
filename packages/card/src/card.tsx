import type {
  BackgroundColorAttributes,
  BoxShadowAttributes,
  HeightAttributes,
  PaddingAttributes,
  TestIdAttributes,
  WidthAttributes,
} from '@waystone/types';
import type { ComponentProps, ReactElement, ReactNode } from 'react';
import { Box } from 'theme-ui';

export type CardImageCss = {
  src?: string;
} & HeightAttributes &
  BackgroundColorAttributes;

const isCardImageCss = (
  image: CardImageCss | ReactElement<ComponentProps<'img'>>
): image is CardImageCss => {
  if ('props' in image) return false;
  if ('type' in image) return false;
  if ('key' in image) return false;

  return true;
};

export type CardProps = {
  /** The components to render inside the card */
  children?: ReactNode;
  /**
   * The image to render inside the card. Provide css options apply the image src as a CSS background image.
   * Alternatively, provide an `img` component to render.
   */
  image?: CardImageCss | ReactElement<ComponentProps<'img'>>;
} & WidthAttributes &
  PaddingAttributes &
  BoxShadowAttributes &
  TestIdAttributes;

export const Card = ({
  image,
  children,
  boxShadow = 'medium',
  'data-testid': testId,
  ...rest
}: CardProps) => {
  const hasImage = !!image;
  const isImageCss = hasImage && isCardImageCss(image);
  const isImageComponent = hasImage && !isImageCss;

  return (
    <Box
      sx={{
        ...rest,
        backgroundColor: 'card',
        borderRadius: 'medium',
        boxShadow,
        overflow: 'clip',
      }}
      data-testid={testId}
    >
      {isImageCss && (
        <div
          sx={{
            backgroundImage: `url(${image.src})`,
            backgroundColor: image.backgroundColor,
            backgroundSize: 'cover',
            backgroundRepeat: 'no-repeat',
            backgroundPosition: 'center',
            height: image.height ?? 'medium',
            maxHeight: image.maxHeight,
            minHeight: image.minHeight,
            width: 'full',
          }}
          aria-description="A decorative background image"
          aria-hidden
        />
      )}
      {isImageComponent && image}
      {children}
    </Box>
  );
};
