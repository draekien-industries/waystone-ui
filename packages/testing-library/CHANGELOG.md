# @waystone/testing-library

## 9.2.0

### Minor Changes

- 21db598: chore: bump deps and add documentation

### Patch Changes

- Updated dependencies [21db598]
  - @waystone/theme@9.3.0

## 9.1.6

### Patch Changes

- 41664e2: fix: use type imports to improve bundling
- Updated dependencies [41664e2]
  - @waystone/theme@9.2.5

## 9.1.5

### Patch Changes

- 856ba2b: fix: missing dependencies when installing package
- Updated dependencies [856ba2b]
  - @waystone/theme@9.2.4

## 9.1.4

### Patch Changes

- 5b726aa: fix build and dependencies
- Updated dependencies [5b726aa]
  - @waystone/theme@9.2.3

## 9.1.3

### Patch Changes

- 7086019: Update dependencies
- Updated dependencies [7086019]
  - @waystone/theme@9.2.2

## 9.1.2

### Patch Changes

- @waystone/theme@9.2.1

## 9.1.1

### Patch Changes

- Updated dependencies [8f75dd6]
  - @waystone/theme@9.2.0

## 9.1.0

### Minor Changes

- 5c43357: Switch to pnpm and implement minimal syntax highlighter

### Patch Changes

- Updated dependencies [5c43357]
  - @waystone/theme@9.1.0

## 9.0.0

### Major Changes

- 4a19c42: Update package structure

  Created the `@waystone/ui` package for easier installation and created the `@waystone/testing-library`
  package for a preconfigured testing-library render function.

  Uninstall your `@waystone/*` packages, and run the command `yarn add @waystone/ui`
  to update to the latest version. You will need to update the imports to point
  to the new source.

### Patch Changes

- Updated dependencies [4a19c42]
  - @waystone/theme@9.0.0
