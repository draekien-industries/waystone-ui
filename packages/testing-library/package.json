{
  "name": "@waystone/testing-library",
  "version": "10.0.2",
  "private": true,
  "main": "./src/index.ts",
  "repository": {
    "type": "git",
    "url": "https://github.com/draekien-industries/waystone-ui"
  },
  "description": "The react component library used by draekien-industries",
  "author": "draekien-industries",
  "sideEffects": false,
  "license": "MIT",
  "files": [
    "src/**"
  ],
  "scripts": {
    "lint": "eslint --ext .tsx,.ts,.jsx,.js ./src/ --fix",
    "lint:release": "eslint --ext .tsx,.ts,.jsx,.js ./src/"
  },
  "devDependencies": {
    "@waystone/tsconfig": "workspace:*",
    "eslint-config-waystone": "workspace:*",
    "eslint-plugin-waystone": "workspace:*",
    "react": "^18.3.1",
    "react-dom": "^18.3.1"
  },
  "dependencies": {
    "@testing-library/jest-dom": "^6.6.3",
    "@testing-library/react": "^16.2.0",
    "@testing-library/user-event": "^14.6.0",
    "@theme-ui/color": "^0.17.1",
    "@theme-ui/css": "^0.17.1",
    "@waystone/theme-provider": "workspace:*",
    "jest": "^29.7.0",
    "jest-environment-jsdom": "^29.7.0",
    "theme-ui": "^0.17.1"
  },
  "publishConfig": {
    "access": "public"
  }
}
