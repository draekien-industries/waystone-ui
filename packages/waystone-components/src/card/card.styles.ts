import { ThemeUIStyleObject } from 'theme-ui';
import { CanBeFullWidth, HasHeight, HasWidth } from '../common/interfaces';

/** THe props for styling the card container */
interface CardContainerCssProps extends HasWidth, HasHeight, CanBeFullWidth {
  /** Should the card not have a shadow */
  noShadow?: boolean;
  /** Is the app in dark mode. */
  darkMode?: boolean;
}

export const cardContainerCss = ({
  fullWidth,
  width,
  minWidth,
  maxWidth,
  noShadow,
  darkMode,
  ...rest
}: CardContainerCssProps): ThemeUIStyleObject => ({
  borderRadius: 'md',
  width: fullWidth ? '100%' : width || 'auto',
  minWidth: !fullWidth && minWidth,
  maxWidth: !fullWidth && maxWidth,
  backgroundColor: darkMode ? 'b-600' : 'b-000',
  boxShadow: !noShadow && 'md',
  overflowY: 'auto',
  ...rest,
});

export const coverImageCss = (coverImageUrl?: string): ThemeUIStyleObject => ({
  backgroundImage: coverImageUrl && `url(${coverImageUrl})`,
  backgroundColor: 'secondary',
  backgroundSize: 'cover',
  backgroundRepeat: 'no-repeat',
  backgroundPosition: 'center',
  width: '100%',
  height: 'md',
});

export const cardContentCss = (noPadding?: boolean): ThemeUIStyleObject => ({
  padding: noPadding ? 0 : 'md',
});
