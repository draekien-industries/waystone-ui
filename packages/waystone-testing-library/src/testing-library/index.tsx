'use client';

import type { RenderOptions, RenderResult } from '@testing-library/react';
import { render } from '@testing-library/react';
import { Waystone } from '@waystone/theme-provider';
import type { PropsWithChildren } from 'react';
import type React from 'react';
import { useEffect } from 'react';
import { useColorMode } from 'theme-ui';

const TestProvider = ({ children }: PropsWithChildren) => (
  <Waystone>{children}</Waystone>
);

const DarkModeWrapper = ({ children }: PropsWithChildren) => {
  const [, setColorMode] = useColorMode();

  useEffect(() => {
    setColorMode('dark');
  }, [setColorMode]);

  return <div>{children}</div>;
};

const themedRender: (
  ui: React.ReactElement,
  options?: RenderOptions
) => RenderResult = (ui, options?) =>
  render(ui, { wrapper: options?.wrapper || TestProvider, ...options });

export * from '@testing-library/react';
export * from '@testing-library/user-event';

export { DarkModeWrapper, themedRender as render };
