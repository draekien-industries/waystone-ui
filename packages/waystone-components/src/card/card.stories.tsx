import React from 'react';
import { Meta, Story } from '@storybook/react';
import { Card, CardProps } from './card';
import exampleImage from './assets/example-image.png';

export default {
  title: 'Components/Card',
  component: Card,
  args: {
    children:
      'Lorem ipsum dolor sit amet consectetur adipisicing elit. Possimus harum dolor nemo atque reiciendis recusandae nam, voluptate earum? Sint, quae omnis! Enim labore harum reiciendis hic id. Eaque, dolores impedit?',
    fullWidth: false,
  },
} as Meta;

const Template: Story<CardProps> = (args) => <Card {...args} />;

export const Default = Template.bind({});

export const WithConstraints = Template.bind({});

WithConstraints.args = {
  maxWidth: 'md',
  maxHeight: 'md',
};

export const WithCoverImage = Template.bind({});

WithCoverImage.args = {
  coverImageUrl: exampleImage,
  maxWidth: 'xl',
};
