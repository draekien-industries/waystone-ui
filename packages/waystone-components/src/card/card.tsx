'use client';

import { useColorMode } from 'theme-ui';
import type { CanBeFullWidth, HasHeight, HasWidth } from '../common/interfaces';
import * as styles from './card.styles';
import type { CoverImageCssOptions } from './card.styles';

/**
 * The props for the Card component.
 */
export type CardProps = {
  /** The children to render in the card. */
  children: React.ReactNode;
  /** Setting this prop to `true` will disable this component's drop shadow. */
  noShadow?: boolean;
  /** Setting this prop to `true` will disable this component's default padding. */
  noPadding?: boolean;
  /** An optional cover image to render inside the card. */
  coverImage?: CoverImageCssOptions;
} & HasWidth &
  HasHeight &
  CanBeFullWidth;

/**
 * A card is a container for content that is grouped together.
 */
export const Card = ({
  children,
  coverImage,
  noPadding,
  ...rest
}: CardProps) => {
  const [mode] = useColorMode();
  const darkMode = mode === 'dark';

  return (
    <div
      sx={styles.cardContainerCss({
        ...rest,
        darkMode,
      })}
    >
      {coverImage && <div sx={styles.coverImageCss(coverImage)} role="img" />}
      <div sx={styles.cardContentCss(noPadding)}>{children}</div>
    </div>
  );
};
