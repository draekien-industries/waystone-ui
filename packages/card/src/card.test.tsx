import { render, screen } from '@waystone/testing-library';
import { Card } from './card';

describe('<Card />', () => {
  it('renders a blank card', () => {
    render(<Card data-testid="test-card" />);

    const card = screen.getByTestId('test-card');

    expect(card).toBeInTheDocument();
  });

  it('renders children', () => {
    render(<Card>lorem ipsum</Card>);

    const card = screen.getByText('lorem ipsum');

    expect(card).toBeInTheDocument();
  });

  it('renders a background image', () => {
    const { container } = render(<Card image={{ src: '/testimage.jpg' }} />);

    // eslint-disable-next-line testing-library/no-container, testing-library/no-node-access
    const backgroundImage = container.querySelector(
      'div[aria-description="A decorative background image"]'
    );

    expect(backgroundImage).toBeInTheDocument();
  });

  it('renders an image component', () => {
    render(
      <Card
        image={
          <img src="/testimage.jpg" alt="test" data-testid="image-component" />
        }
      />
    );

    const image = screen.getByTestId('image-component');

    expect(image).toBeInTheDocument();
  });
});
