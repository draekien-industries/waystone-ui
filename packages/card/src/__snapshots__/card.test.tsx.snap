// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Card /> renders a background image 1`] = `
<div>
  <div
    class="css-1aeqb5q"
  >
    <div
      aria-description="A decorative background image"
      aria-hidden="true"
      class="css-hfn5ud"
    />
  </div>
</div>
`;

exports[`<Card /> renders a blank card 1`] = `
<div>
  <div
    class="css-1aeqb5q"
    data-testid="test-card"
  />
</div>
`;

exports[`<Card /> renders an image component 1`] = `
<div>
  <div
    class="css-1aeqb5q"
  >
    <img
      alt="test"
      data-testid="image-component"
      src="/testimage.jpg"
    />
  </div>
</div>
`;

exports[`<Card /> renders children 1`] = `
<div>
  <div
    class="css-1aeqb5q"
  >
    lorem ipsum
  </div>
</div>
`;
